### PR TITLE
JSSDK-97 - Update clients to match new access control

### LINKED ISSUE #97: Update clients with new access control
Update things to match the new access control.
This will include:

- [ ] adding the roles, role-actions and role-members methods
- [ ] Updating documentation
- [ ] Adding tests
- [ ] Updating the other methods as needed

Related to https://github.com/absmach/magistrala-ui-new/issues/293

### PR DESCRIPTION
**Pull request by wambui-pixel**
Pull request title should be `JSSDK-XXX - description` or `NOISSUE - description` where XXX is ID of issue that this PR relate to.
Please review the [CONTRIBUTING.md](./CONTRIBUTING.md) file for detailed contributing guidelines.

### What does this do?
This updates clients service to match new auth
### Which issue(s) does this PR fix/relate to?
Resolves #97 
Put here `Resolves #XXX` to auto-close the issue that your PR fixes (if such)

### List any changes that modify/break current functionality
n/a
### Have you included tests for your changes?
yes
### Did you document any new/modified functionality?
yes

### CODE REVIEW

**Review by dborovcanin — APPROVED** [review]

File: examples/clients.ts [hunk 1]
```diff
@@ -0,0 +1,293 @@
+// Import the SDK class from the mainflux-sdk package
```
Comment (ianmuchyri) [inline examples/clients.ts:1]: remove this comment.

File: examples/clients.ts [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+import SDK from "../src/sdk";
+
+const defaultUrl = "http://localhost";
+
+const mySdk = new SDK({
+  clientsUrl: `${defaultUrl}:9006`,
+});
+
+// Clients.ts examples.
```
Comment (ianmuchyri) [inline examples/clients.ts:10]: remove this comment

File: src/clients.ts [hunk 3]
```diff
@@ -0,0 +1,974 @@
...
+      const ClientData: ClientsPage = await response.json();
+      return ClientData;
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  /**
+  * @method Enable - Enables a previously disabled client by its ID.
+  * @param {string} clientID - The  unique ID of the client.
```
Comment (ianmuchyri) [inline src/clients.ts:126]:
```suggestion
  * @param {string} clientId - The  unique ID of the client.
```

File: src/clients.ts [hunk 4]
```diff
@@ -0,0 +1,974 @@
...
+      const clientData: Client = await response.json();
+      return clientData;
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  /**
+  * @method UpdateClientTags - Updates an existing client's tags.
+  * @param {client} Client - Client object with updated properties.
```
Comment (ianmuchyri) [inline src/clients.ts:285]:
```suggestion
  * @param {Client} client - Client object with updated properties.
```

File: src/clients.ts [hunk 5]
```diff
@@ -0,0 +1,974 @@
...
+      return clientsData;
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  /**
+  * @method ClientParents - Sets parent to a client.
+  * @param {string} domainId - The  unique ID of the domain.
+  * @param {string} clientlId - The unique ID of the client to be updated.
```
Comment (ianmuchyri) [inline src/clients.ts:459]:
```suggestion
  * @param {string} clientId - The unique ID of the client to be updated.
```

File: src/clients.ts [hunk 6]
```diff
@@ -0,0 +1,974 @@
...
+  * @param {string} parentGroupId - The unique ID of the group to be set as the parent.
+  * @param {string} token - Authorization token.
+  * @returns {Promise<Response>} - A promise that resolves when the parent group is successfully set for the specified client.
+  * @throws {Error} If the parent group cannot be set for the client.
+  */
+  public async ClientParents(domainId: string, clientlId: string, parentGroupId: string, token: string) : Promise<Response> {
```
Comment (ianmuchyri) [inline src/clients.ts:465]:
```suggestion
  public async ClientParents(domainId: string, clientId: string, parentGroupId: string, token: string) : Promise<Response> {
```

File: src/clients.ts [hunk 7]
```diff
@@ -0,0 +1,974 @@
...
+      method: "POST",
+      headers: {
+        "Content-Type": this.contentType,
+        Authorization: `Bearer ${token}`
+      },
+      body: JSON.stringify({ parent_group_id: parentGroupId })
+    };
+    try {
+      const response = await fetch(
+        new URL(`${domainId}/${this.clientsEndpoint}/${clientlId}/parent`, this.clientsUrl).toString(),
```
Comment (ianmuchyri) [inline src/clients.ts:476]:
```suggestion
        new URL(`${domainId}/${this.clientsEndpoint}/${clientId}/parent`, this.clientsUrl).toString(),
```

File: src/clients.ts [hunk 8]
```diff
@@ -0,0 +1,974 @@
...
+  /**
+  * @method DeleteClient - Deletes a client with specified id..
+  * @param {string} clientId - The  unique ID of the client.
+  * @param {string} domainId - The  unique ID of the domain.
+  * @param {string} token - Authorization token.
+  * @returns {Promise<Response>} - A promise that resolves when the client is deleted.
+  * @throws {Error} If the client cannot be deleted.
+  */
+  public async DeleteClient(
+    ClientId: string,
```
Comment (ianmuchyri) [inline src/clients.ts:534]:
```suggestion
    clientId: string,
```

File: src/clients.ts [hunk 9]
```diff
@@ -0,0 +1,974 @@
...
+      headers: {
+        "Content-Type": this.contentType,
+        Authorization: `Bearer ${token}`,
+      },
+    };
+
+    try {
+      const response = await fetch(
+        new URL(
+          `${domainId}/${this.clientsEndpoint}/${ClientId}`,
```
Comment (ianmuchyri) [inline src/clients.ts:549]:
```suggestion
          `${domainId}/${this.clientsEndpoint}/${clientId}`,
```

File: src/clients.ts [hunk 10]
```diff
@@ -0,0 +1,974 @@
...
+      );
+      return role;
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  /**
+   * @method ListClientRoles - Lists all roles within a specific client.
+   * @param {string} clientId - The unique identifier of the client.
```
Comment (ianmuchyri) [inline src/clients.ts:624]: you are missing `domainId`

File: src/clients.ts [hunk 11]
```diff
@@ -0,0 +1,974 @@
...
+      );
+      return rolesPage;
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  /**
+   * @method ViewClientRole - Retrieves details about a specific role in a client.
+   * @param {string} clientId - The unique identifier of the client.
```
Comment (ianmuchyri) [inline src/clients.ts:652]: you are missing `domainId`

File: src/clients.ts [hunk 12]
```diff
@@ -0,0 +1,987 @@
...
+      }
+      const enabledClient: Client = await response.json();
+      return enabledClient;
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  /**
+   * @method Disable - Disables an enabled client by its ID..
```
Comment (ianmuchyri) [inline src/clients.ts:165]:
```suggestion
   * @method Disable - Disables an enabled client by its ID.
```

File: src/clients.ts [hunk 13]
```diff
@@ -0,0 +1,987 @@
...
+  /**
+  * @method ClientParents - Sets parent to a client.
+  * @param {string} domainId - The  unique ID of the domain.
+  * @param {string} clientId - The unique ID of the client to be updated.
+  * @param {string} parentGroupId - The unique ID of the group to be set as the parent.
+  * @param {string} token - Authorization token.
+  * @returns {Promise<Response>} response - A promise that resolves when the parent group is successfully set for the specified client.
+  * @throws {Error} - If the parent group cannot be set for the client.
+  */
+  public async ClientParents(domainId: string, clientId: string, parentGroupId: string, token: string) : Promise<Response> {
```
Comment (ianmuchyri) [inline src/clients.ts:465]:
```suggestion
  public async SetClientParentGroup(domainId: string, clientId: string, parentGroupId: string, token: string) : Promise<Response> {
```

File: src/clients.ts [hunk 14]
```diff
@@ -0,0 +1,987 @@
...
+
+  /**
+  * @method DeleteClientParents - Removes the parent group from a specified client.
+  * @param {string} domainId - The  unique ID of the domain.
+  * @param {string} clientId - The  unique ID of the client.
+  * @param {string} token - Authorization token.
+  * @returns {Promise<Response>} response - A promise that resolves when the parent group is successfully removed from the specified client.
+  * @throws {Error} - If the parent group cannot removed from the client.
+  */
+  public async DeleteClientParents(domainId: string, clientId: string, token: string) : Promise<Response> {
```
Comment (ianmuchyri) [inline src/clients.ts:498]:
```suggestion
  public async DeleteClientParentGroup(domainId: string, clientId: string, token: string) : Promise<Response> {
```

File: src/clients.ts [hunk 15]
```diff
@@ -0,0 +1,987 @@
...
+      }
+      const deleteClientParentsResponse: Response = { status: response.status, message: "Client group parent deleted successfully" };
+      return deleteClientParentsResponse;
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  /**
+  * @method DeleteClient - Deletes a client with specified id..
```
Comment (ianmuchyri) [inline src/clients.ts:526]:
```suggestion
  * @method DeleteClient - Deletes a client with specified id.
```

File: src/defs.ts [hunk 16]
```diff
@@ -30,30 +30,32 @@ export interface Credentials {
   secret?: string;
 }
 
-export interface ThingCredentials {
+export interface ClientCredentials {
```
Comment (ianmuchyri) [inline src/defs.ts:33]: let's call the other `Credentials` `UserCredentials`

File: src/defs.ts [hunk 17]
```diff
@@ -91,16 +93,13 @@ export interface ChannelBasicInfo {
...
   domain_id?: string | DomainBasicInfo;
   metadata?: Record<string, any>;
   tags?: string[];
-  ParentGroup?: string;
-  level?: number;
-  path?: string;
+  parent_group_id?: string;
   created_at?: Date;
   updated_at?: Date;
   updated_by?: string;
```
Comment (ianmuchyri) [inline src/defs.ts:105]: It also has permissions:
https://github.com/absmach/magistrala/blob/9bf293fd44bbeceb7ac6409c7cd12bb9caac8f74/channels/channels.go#L29

File: src/defs.ts [hunk 18]
```diff
@@ -291,8 +290,7 @@ export interface BootstrapConfig {
   channels?: string[];
   external_id?: string;
   external_key?: string;
-  thing_id?: string;
-  thing_key?: string;
+  client_id?: string;
```
Comment (ianmuchyri) [inline src/defs.ts:293]:
```suggestion
  client_id?: string;
  client_secret?: string
```

**Review by ianmuchyri — CHANGES_REQUESTED** [review]

File: src/defs.ts (hunk 17 above)
Comment (wambui-pixel) [reply src/defs.ts:105]: https://github.com/absmach/sdk-js/blob/c39b9c2590b24adb3484c53216eca0301c1b4c5c/src/defs.ts#L106

**Review by ianmuchyri — APPROVED** [review]

**Pull request merged**